### DISCUSSION TITLE
Swallow exceptions from ObserveDurationAsync, but throw from Func input?

### DISCUSSION THREAD

**Discussion by cscoon** [PrometheusClientNet/Prometheus.Client#182] [Q&A] [answered]
We wrap our Prometheus metric emitting calls in a wrapper class which essentially catches any exceptions thrown while incrementing a metric and swallows it.  The reason is that if we are performing an important business process, we would rather ignore the exception from `IncTo` than have it throw an exception and end the business process.

This works great for most Prometheus.Client methods such as `IncTo`.  But the practice breaks with `ObserveDurationAsync` because that method accepts the code snippet that should be timed as a `Func`.  So both pieces of code are executed in a single call.

What we'd like to do is ignore/swallow any exception that results from `ObserveDurationAsync`, but throw any exception that results from the `Func` execution.

Does Prometheus.Client provide a pattern for doing this?

**Comment by sanych-sun** [comment, marked as answer]
No. We do not have anything like this for now, but probably can add this as a extension method argument or even metric config. The first approach allow us to localize the changes, but last one looks more generic. But not sure if it makes sense.

@phnx47 what do you think? Is adding new metric family configuration (for any kind of metric metrics) to ignore any exceptions while observations is valuable? I mean suppress any exception that could appear inside Inc, Dec, etc methods. If we go this way we probably have to at least report the exception to diagnostics source.

**Reply by phnx47** [reply]
I didn't fully understand the current problem. `ObserveDuration` and `ObserveDurationAsync` just extensions, so can implement other extensions.

@cscoon Maybe you can create PR with your suggestions?

**Reply by phnx47** [reply]
Issue here: https://github.com/prom-client-net/prom-client/issues/183
